### PR TITLE
[Notifier] add support for gatewayapi-notifier

### PR DESCRIPTION
**Pull request by PGLongo**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | [symfony-docs/pull/14463](https://github.com/symfony/symfony-docs/pull/14463)

I've added support to [GatewayAPI](https://gatewayapi.com/) with Notifier

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Piergiuseppe Longo <piergiuseppe.longo@gmail.com>
+ * @experimental in 5.2
+ */
+final class GatewayAPITransport extends AbstractTransport
+{
+    protected const HOST = 'gatewayapi.com';
```
Comment (Nyholm) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: Why protected?

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php [hunk 2]
```diff
@@ -15,7 +15,6 @@
 use Doctrine\DBAL\Driver\AbstractMySQLDriver;
 use Doctrine\DBAL\DriverManager;
 use Doctrine\DBAL\Schema\Schema;
-use Doctrine\DBAL\Version;
```
Comment (Nyholm) [inline src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php:18]: I know your fingers are itchy, but try to not do unrelated CS fixes. 

(You dont need to revert this. It was just an FYI)

**Review by Nyholm — COMMENTED** [review]
Wow. Thank you. What a great first contribution. 

Could I ask you to add a some tests to this feature too?

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransportFactory.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+
+/**
+ * @author Piergiuseppe Longo <piergiuseppe.longo@gmail.com>
+ * @experimental in 5.2
+ */
+class GatewayAPITransportFactory extends AbstractTransportFactory
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransportFactory.php]: Final?

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+===============
+
+Provides GatewayAPI integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+// .env file
+GATEWAYAPI_DSN=gatewaypi://TOKEN@default?from=FROM
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md]: missing "a" gateway**a**pi

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    private $authToken;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: I am not sure, do other notifier use property types?

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 1 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: I don't know. It is protected in other Transports, like: FreeMobileTransport, SmsapiTransport, NexmoTransport, etc.

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php (hunk 2 above)
Comment (PGLongo) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php:18]: Sorry, my mistake with php-cs-fixer.

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransportFactory.php (hunk 3 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md (hunk 4 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 5 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: You are talking about hosts? If yes, it's declared in other transports like FreeMobileTransport, SmsapiTransport, NexmoTransport, etc.

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: No I mean if they use `private string $authToken` or not

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -2227,6 +2228,7 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             SendinblueNotifierTransportFactory::class => 'notifier.transport_factory.sendinblue',
             DiscordTransportFactory::class => 'notifier.transport_factory.discord',
             LinkedInTransportFactory::class => 'notifier.transport_factory.linkedin',
+            GatewayAPITransportFactory::class => 'notifier.transport_factory.gatewayapi',
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I would prefer:
```diff
- GatewayAPITransportFactory
+ GatewayApiTransportFactory
```

Same for all the others, but lets wait for @fabpot meaning on this

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $this->authToken = $authToken;
+        $this->from = $from;
+
+        parent::__construct($client, $dispatcher);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('gatewaypi://%s?from=%s', $this->getEndpoint(), $this->from);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]:
```suggestion
        return sprintf('gatewayapi://%s?from=%s', $this->getEndpoint(), $this->from);
```

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
+GatewayAPI Notifier
+===============
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md]:
```suggestion
GatewayAPI Notifier
===================
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (PGLongo) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: The official name of the service is GatewayAPI, I prefer GatewayApiTransportFactory too.

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 5 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: Some Transport use it, for instance, `SmsapiTransport`.

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md (hunk 8 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/README.md]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 7 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: Done

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I agree with Oskar, let's be consistent with the rest of the code base.

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 1 above)
Comment (Nyholm) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: Let's make it private then =)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (PGLongo) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Done the refactor. I have maintained GatewayAPI only in `composer.json`, `README.md` and `phpunit.xml.dist` when the official name is provided, like:

`<testsuite name="Symfony GatewayAPI Notifier Bridge Test Suite">`

**Comment by PGLongo** [comment]
@Nyholm I will add some tests

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 1 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: I don't agree with you. The host is protected in the other Transports and `AbstractTransport`, so it would be better to keep it protected to be consistent.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (OskarStark) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Thank you 👍

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (PGLongo) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Thank you @OskarStark ❤️

**Comment by PGLongo** [comment]
@Nyholm, are the test ok?

**Comment by PGLongo** [comment]
I've forgotten to mention that I've added the [recipe](https://github.com/symfony/recipes/pull/835)

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    "name": "symfony/gatewayapi-notifier",
+    "type": "symfony-bridge",
+    "description": "Symfony GatewayAPI Notifier Bridge",
+    "keywords": ["sms", "gatewayapi", "notifier"],
+    "homepage": "https://gatewayapi.com",
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "Piergiuseppe Longo",
+            "email": "piergiuseppe.longo@gmail.com"
```
Comment (Nyholm) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json:15]:
```suggestion
            "name": "Fabien Potencier",
            "email": "fabien@symfony.com"
```

**Review by Nyholm — COMMENTED** [review]
Im happy with this after one small change. 

Thank you

**Comment by PGLongo** [comment]
Thank you very much! I'm so happy to contribute to Symfony after years of usage.

**Review by Nyholm — APPROVED** [review]
Thank you. 

Could you rebase this so it shows that the tests are green?

**Comment by PGLongo** [comment]
I'm sorry @Nyholm but I have done the rebase twice but the tests are not passing due to other problems.

**Comment by Nyholm** [comment]
That is fine. Thank you. 

There are a few of us that have been trying to fix the tests now and then for the past week. There are too many merges. =)

There is nothing more you need to do atm.

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/.gitattributes [hunk 10]
```diff
@@ -0,0 +1,3 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitignore export-ignore
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/.gitattributes:4]: Missing `/.gitattributes export-ignore`

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SentMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Piergiuseppe Longo <piergiuseppe.longo@gmail.com>
+ * @experimental in 5.2
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]: Should be 5.3 now.
Can you also add a blank line just above?
The same goes for all other classes.

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    "autoload": {
+        "psr-4": {
+            "Symfony\\Component\\Notifier\\Bridge\\GatewayApi\\": ""
+        },
+        "exclude-from-classmap": [
+            "/Tests/"
+        ]
+    },
+    "minimum-stability": "dev",
+    "extra": {
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]: Can you remove the `extra` section as we don't need it anymore.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by fabpot — CHANGES_REQUESTED** [review]
Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php (hunk 11 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json (hunk 12 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]: Done

**Comment by PGLongo** [comment]
> Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
+GatewayAPI Notifier
+===================
+
+Provides GatewayAPI integration for Symfony Notifier.
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md]:
```suggestion
Provides [GatewayAPI](https://gatewayapi.com) integration for Symfony Notifier.
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+DSN example
+-----------
+
+```
+// .env file
+GATEWAYAPI_DSN=gatewayapi://TOKEN@default?from=FROM
+```
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md:12]:
```suggestion
```ini
gatewayapi://TOKEN@default?from=FROM
```
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $transport = $this->getTransport();
+        $message = new ChatMessage('3333333333');
+        $this->assertFalse($transport->supports($message));
+    }
+
+    private function getTransport(MockHttpClient $client = null): GatewayApiTransport
+    {
+        if (null == $client) {
+            $this->createMock(ResponseInterface::class);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Looks like this is not used 🤔

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    "authors": [
+        {
+            "name": "Fabien Potencier",
+            "email": "fabien@symfony.com"
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]: You should add yourself to the authors

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md [hunk 17]
```diff
@@ -0,0 +1,26 @@
+GatewayAPI Notifier
+===================
+
+Provides GatewayAPI integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+// .env file
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/README.md]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Notifier\Bridge\GatewayApi\GatewayApiTransportFactory;
+use Symfony\Component\Notifier\Exception\IncompleteDsnException;
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\Dsn;
+
+final class GatewayApiTransportFactoryTest extends TestCase
+{
+    public function testSupportsGatewayApiScheme(): void
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]: Please remove `:void` from all test methods

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HttpClient\MockHttpClient;
+use Symfony\Component\Notifier\Bridge\GatewayApi\GatewayApiTransport;
+use Symfony\Component\Notifier\Message\ChatMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+final class GatewayApiTransportTest extends TestCase
+{
+    public function testSend(): void
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Please remove `:void` from all test methods

File: src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php (hunk 5 above)
Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/GatewayAPI/GatewayAPITransport.php]: Symfony 5.x supports PHP 7.2+ so we cannot use typed properties.

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php (hunk 15 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Yes, it is used 3 times.

image

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Hmm but this mock is neither returned nor assigned to any variable 🧐

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+    public function testCreateWithNoFromShouldThrowsIncompleteDsnException()
+    {
+        $factory = $this->initFactory();
+        $this->expectException(IncompleteDsnException::class);
+        $dsn = 'gatewayapi://token@default';
+        $factory->create(Dsn::fromString($dsn));
+    }
+
+    private function initFactory(): GatewayApiTransportFactory
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]: Can you please rename this method to createFactory() ?

Thanks

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php (hunk 15 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Sorry, you are right. I was thinking you were talking about `getTransport`

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]: Thanks

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/.gitattributes [hunk 21]
```diff
@@ -0,0 +1,4 @@
...
+/.gitattributes export-ignore
```
Comment (jderusse) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/.gitattributes:3]: tiny detail: Please use same order as all other component
```
/Tests export-ignore
/phpunit.xml.dist export-ignore
/.gitattributes export-ignore
/.gitignore export-ignore
```

Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/.gitattributes:3]: Thank you @jderusse.

**Review by fabpot — APPROVED** [review]
PLEASE don't merge yet. I will merge all similar PRs in one go (as I need to handle the subtree splits, recipes, ... much faster to do everything at the same time).

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof SmsMessage;
+    }
+
+    protected function doSend(MessageInterface $message): SentMessage
+    {
+        if (!$message instanceof SmsMessage) {
+            throw new LogicException(sprintf('The "%s" transport only supports instances of "%s" (instance of "%s" given).', __CLASS__, SmsMessage::class, get_debug_type($message)));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]: Please use newly introduced `UnsupportedMessageTypeException` here.

Fore reference: https://github.com/symfony/symfony/pull/39365

Thanks

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+        if (!$authToken) {
+            throw new IncompleteDsnException('Missing auth token.', $dsn->getOriginalDsn());
+        }
+
+        if ('gatewayapi' === $scheme) {
+            return (new GatewayApiTransport($authToken, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
+        }
+
+        throw new UnsupportedSchemeException($dsn, 'gatewayapi', $this->getSupportedSchemes());
+    }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]: Please pull up the scheme check like here, to first check if the scheme is supported and afterwards for missing options. Thanks

https://github.com/symfony/symfony/blob/7d846d3e82f4e3dabec7d4b5300f50a1dd9ed7d6/src/Symfony/Component/Notifier/Bridge/Sinch/SinchTransportFactory.php#L27-L47

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+final class GatewayApiTransportFactoryTest extends TestCase
+{
+    public function testSupportsGatewayApiScheme()
+    {
+        $factory = $this->createFactory();
+
+        $dsn = 'gatewayapi://token@default?from=Symfony';
+
+        $this->assertTrue($factory->supports(Dsn::fromString($dsn)));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $this->assertTrue($factory->supports(Dsn::fromString('gatewayapi://token@host.test?from=Symfony')));
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+        $dsn = 'gatewayapi://default?from=Symfony';
+        $factory->create(Dsn::fromString($dsn));
+    }
+
+    public function testCreateWithNoFromShouldThrowsIncompleteDsnException()
+    {
+        $factory = $this->createFactory();
+        $this->expectException(IncompleteDsnException::class);
+        $dsn = 'gatewayapi://token@default';
+        $factory->create(Dsn::fromString($dsn));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $factory = $this->createFactory();

        $this->expectException(IncompleteDsnException::class);

        $factory->create(Dsn::fromString('gatewayapi://token@host.test'));
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+        $dsn = 'wrongGateway://token@default?from=Symfony';
+        $factory->create(Dsn::fromString($dsn));
+    }
+
+    public function testCreateWithNoTokenThrowsIncompleteDsnException()
+    {
+        $factory = $this->createFactory();
+        $this->expectException(IncompleteDsnException::class);
+        $dsn = 'gatewayapi://default?from=Symfony';
+        $factory->create(Dsn::fromString($dsn));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $factory = $this->createFactory();

        $this->expectException(IncompleteDsnException::class);

        $factory->create(Dsn::fromString('gatewayapi://host.test?from=Symfony'));
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+
+        $this->assertTrue($factory->supports(Dsn::fromString($dsn)));
+    }
+
+    public function testUnSupportedGatewayShouldThrowsUnsupportedSchemeException()
+    {
+        $factory = $this->createFactory();
+        $this->expectException(UnsupportedSchemeException::class);
+        $dsn = 'wrongGateway://token@default?from=Symfony';
+        $factory->create(Dsn::fromString($dsn));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $factory = $this->createFactory();

        $this->expectException(UnsupportedSchemeException::class);

        $factory->create(Dsn::fromString( 'somethingElse://token@default?from=Symfony'));
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+        $this->assertFalse($transport->supports($message));
+    }
+
+    private function getTransport(MockHttpClient $client = null): GatewayApiTransport
+    {
+        return (new GatewayApiTransport(
+            'authtoken',
+            'Symfony',
+            $client
+        ))->setHost('host.test');
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]:
```suggestion
        return (new GatewayApiTransport('authtoken', 'Symfony', $client))->setHost('host.test');
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "^5.2"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]:
```suggestion
        "symfony/notifier": "^5.3"
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+    },
+    "autoload": {
+        "psr-4": {
+            "Symfony\\Component\\Notifier\\Bridge\\GatewayApi\\": ""
+        },
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]:
```suggestion
        "psr-4": { "Symfony\\Component\\Notifier\\Bridge\\GatewayApi\\": "" },
```

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php (hunk 22 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php (hunk 23 above)
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]:
```suggestion
    public function create(Dsn $dsn): TransportInterface
    {
        $scheme = $dsn->getScheme();
        
        if ('gatewayapi' !== $scheme) {
            throw new UnsupportedSchemeException($dsn, 'gatewayapi', $this->getSupportedSchemes());
        }
        
        $authToken = $this->getUser($dsn);
        $from = $dsn->getOption('from');

        if (!$from) {
            throw new IncompleteDsnException('Missing from.', $dsn->getOriginalDsn());
        }

        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
        $port = $dsn->getPort();

        return (new GatewayApiTransport($authToken, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
    }
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+
+        if (!$authToken) {
+            throw new IncompleteDsnException('Missing auth token.', $dsn->getOriginalDsn());
+        }
+
+        if ('gatewayapi' === $scheme) {
+            return (new GatewayApiTransport($authToken, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
+        }
+
+        throw new UnsupportedSchemeException($dsn, 'gatewayapi', $this->getSupportedSchemes());
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]:
```suggestion
        return (new GatewayApiTransport($authToken, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php (hunk 23 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php (hunk 31 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php (hunk 23 above)
Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+        $this->assertTrue($factory->supports(Dsn::fromString('gatewayapi://token@host.test?from=Symfony')));
+    }
+
+    public function testUnSupportedGatewayShouldThrowsUnsupportedSchemeException()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(UnsupportedSchemeException::class);
+
+        $factory->create(Dsn::fromString( 'somethingElse://token@default?from=Symfony'));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $factory->create(Dsn::fromString('somethingElse://token@default?from=Symfony'));
```

**Review by OskarStark — APPROVED** [review]
Looks good to me @fabpot 👍

**Comment by fabpot** [comment]
@PGLongo Can you rebase to get rid of the merge commit? Thank you.

**Comment by PGLongo** [comment]
@fabpot I will try this weekend. I have a lot of conflicts and I don't have the time to fix them right now.

```
git rebase upstream/5.x 
```

```git rebase upstream/5.x 
Auto-merging src/Symfony/Component/Notifier/Transport.php
CONFLICT (content): Merge conflict in src/Symfony/Component/Notifier/Transport.php
Auto-merging src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.php
CONFLICT (content): Merge conflict in src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.php
Auto-merging src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
CONFLICT (content): Merge conflict in src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
error: could not apply 2575ca2b12... [Notifier] add support for gatewayapi-notifier
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 2575ca2b12... [Notifier] add support for gatewayapi-notifier```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "^5.3"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]: should use `~5.3.0`

Comment (PGLongo) [reply src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]: Done

**Comment by PGLongo** [comment]
Please can someone help me with the rebase to get rid of the merge commit as asked by @fabpot? @Nyholm , @nicolas-grekas, or @OskarStark is it correct to have a lot of merge conflicts doing:

`git rebase upstream/5.x `

Have I to do something else?

**Comment by OskarStark** [comment]
Sure, can you please invite me to your repo?
